### ISSUE TITLE
Improving the new Citation System

### ISSUE THREAD

**Issue by PaulHeggarty** [lingdb/CoBL#344]
So that we can have a better view of progress, using check-boxes here, I’m starting a new issue, following on from #306.
- [ ] All three counts in the `Cognacy`, `Cogset` and `Lexeme` columns appear as blue and clickable, but all give a bad gateway error when clicked (for me on Firefox, at least).

- [ ] All author/editor fields need to be split (it seems they are not yet) into separate fields for `author surname` and `author given names`.  They currently appear as just surname [comma] given names or abbreviations (inconsistently, sometimes even with the same author).  Proper systems enter the full first names wherever available, and work out the initials themselves when needed.  Bibtex itself may not do this, but most good bibliography management software does, it is absolutely standard practice now on webpages with bibliographies.  We need to do it too, there is no excuse not to.  See screenshots below from WALS, showing how it stores these as separate fields, and how they correctly get picked up by Zotero.
![screenshot - 2016-11-07 09_06_41](https://cloud.githubusercontent.com/assets/8819910/20050524/1af981d6-a4ca-11e6-8923-fb55c37e8cdc.png)
![screenshot - 2016-11-07 09_06_15](https://cloud.githubusercontent.com/assets/8819910/20050529/1e8d8de2-a4ca-11e6-9224-f21a0eabff60.png)

**Comment by khoidt** [comment]
### Bad gateway
I'm getting the error for _most_ of the links, but not for all of them. @PaulHeggarty, just to see that we are on the same page here, please, check if the following links open:
http://cobl.info/sources/72/cogset/
http://cobl.info/sources/337/lexeme/
I still don't understand what causes it. @runjak, could you send me the server logs for the error?

### Filters
The filters are planned for the next release (by the end of this week the latest). It is a big task, which would take me at least a couple of working days to implement. 

### Names
Storing the name parts separately in different fields on the database level means that in the case of multiple authors or editors one field would have all surnames, while the other would have all given names. Such a data structure seems unnecessary complex with no clear benefits (see below) and will require additional operations for putting the names back together, which means more code, more database queries, more load on the server and a greater potential for errors.
A clear distinction between surnames and given names (which can give the `xml` output on the screenshot above) is easily achieved by strictly following an unambiguous format in the data, e.g.:

surname`, `givenName(` `givenName)(` and `surname`, `givenName)
As in "Smith, John Robert and Doe, Jane Mary".

@nitsnatsnok, would it be possible to format all the names like this in the `.bib` data?

**Comment by nitsnatsnok** [comment]
### Bad gateway
No errors for both of your links on my side, the second one takes a while though.

### Names
That is exactly what I was going to propose. BibTeX splits names up automatically, so that `Doe, Jane Mary` and `Jane Mary Doe` in the `.bib` data is rendered in exactly the same way in the end – as we don't use BibTeX now, we need to adjust the data ourselves. I will try to get this done until next weekend.

**Comment by runjak** [comment]
@khoidt both links work for me right now.
That is with cobl.info not on my machine.
Pages look like this:
![screenshot_2016-11-07_11-36-37](https://cloud.githubusercontent.com/assets/5417642/20054757/7de0f856-a4de-11e6-893b-144ef1633b9b.png)
![screenshot_2016-11-07_11-36-44](https://cloud.githubusercontent.com/assets/5417642/20054758/7de22d52-a4de-11e6-9087-93be35a4981f.png)

Looking at the server logs it appears that some workers have experienced timeouts with page loads taking too long:
![screenshot_2016-11-07_11-37-46](https://cloud.githubusercontent.com/assets/5417642/20054800/ac58e838-a4de-11e6-9cda-0db57eca6776.png)

Any Idea what this could be about?
It may help to investigate this with the aid of the Django Debug Toolbar.
I can try to reproduce on my machine provided you can give me hints as to what you're doing.

**Comment by runjak** [comment]
From looking at the debug toolbar data for the `/sources/337/lexeme/` case I see that this case alone shouldn't be the problem (there are certainly tables that are worse in terms of loading time):
![screenshot_2016-11-07_11-50-50](https://cloud.githubusercontent.com/assets/5417642/20055302/b0023b2c-a4e0-11e6-92b8-459def3a6e42.png)

**Comment by nitsnatsnok** [comment]
It works for all pages up to four entries, pages with more entries consistently yield the timeout error.

**Comment by runjak** [comment]
Thanks - I can reproduce the problem, I think.

**Comment by khoidt** [comment]
@runjak, thanks for the screenshorts! I was using Django's generic `inlineformset_factory` to produce the forms for the views. The seem to work very slow indeed (I also received a `timeout error` a couple of times while testing, but considered it a local issue, as on my machine CoBL generally runs much slower than the remote server).
I believe that the problem is somewhere on the database queries level, but I still cannot get where exactly and why.

**Comment by runjak** [comment]
Yes - the form is fetching the lexemes one by one instead of doing a select over all the wanted cases. This can be mitigated with the use of `prefetch_related` usually, but appears to be a bit more involved than I expeceted in some cases. Compare this [StackOverflow](https://stackoverflow.com/q/9583096/448591) question.

**Comment by runjak** [comment]
Hm - I cannot easily get the `prefetch_related` or other approaches to work.
```
    data = Source.objects.prefetch_related(
        'lexemecitation_set',
        'lexemecitation_set__lexeme',
        'lexeme_set').get(pk=source_id)
```
Still ends up with django fetching each entry separately and even replacing the initial content of the form fields like this:
```
    for lexemecitation, form in izip(data.lexemecitation_set.all(), formset.forms):
        form['comment'].initial = lexemecitation.comment
        form['lexeme'].initial = lexemecitation.lexeme
```
doesn't prevent this from happening.

For now I'll focus on something else - because there are other pressing things I need to attend to - maybe you, @khoidt have luck with something else concerning this problem .)

**Comment by khoidt** [comment]
@runjak, I'm trying to solve this myself right now (also with no significant progress so far). I will let you know when it is done.

**Comment by nitsnatsnok** [comment]
@khoidt Please bear in mind that this is not the only malfunction of the sources table: Since the last change, the “Import” button as well as the “deprecated” column and the “Edit” buttons are missing. Interestingly, all of those seem to be displayed shortly when loading the page.

**Comment by khoidt** [comment]
@nitsnatsnok, this is actually correct preview behavior (except for the short display on loading). These are now hidden from for non-superusers, as it was required in #306.

**Comment by nitsnatsnok** [comment]
@khoidt Please make me a superuser then. :wink:

**Comment by khoidt** [comment]
Sorry, I cannot do it, as I'm not a superuser there myself (only of my own castle, i. e. my locally running version). @runjak, can you?

**Comment by runjak** [comment]
I think this is something @PaulHeggarty and/or @cormacanderson need to decide.
I don't have superuser rights on the site myself, but can gain and grant them easily because I can directly access the database.
I suspect that it would be helpful to have a more fine grained access policy in the future - possibly based on groups rather than direct user flags.

**Comment by khoidt** [comment]
@runjak I'm totally with you about the groups. It is a much more convenient way to specify different rights. If @PaulHeggarty and @cormacanderson agree, this should become a new issue. However, it is not something that can be done very quickly, so the problem of @nitsnatsnok being unable to change the bibliography on the server remains. Alternatively, I can remove the restrictions for non-superusers.

**Comment by runjak** [comment]
Both of them are in a meeting rn, but I'll bug them about this once they're out of it again.

**Comment by PaulHeggarty** [comment]
"Alternatively, I can remove the restrictions for non-superusers."
Please do remove the restriction on _this_ task, yes.

**Comment by PaulHeggarty** [comment]
Generally, this issue is to be dealt with as far as possible by Ilya, because Jakob has many other very high priority issues.  
Also, the order of priority within the three issues here is definitely about how fast each can be done.  It is not so urgent to have the complex third check-box issue fixed.  Please focus on check-boxes 1 and 2 and deploy them as soon as possible, rather than delaying those until the complex one (3) is fixed.  This is a general policy:  within a given priority rank, always try to do and deploy quick things first, so that others can get on with using them, and leave the slower tasks for after the quicker ones are fixed.

**Comment by m-scarborough** [comment]
General question related to the new citation system makeover (this seems as good a place as any to ask):  Is it currently possible to add new sources into the database at the moment, or must these all now be (for now) manually imported through a master BibTeX file?  I'm asking because I want to add a citation to something that's not currently in the sources and I can't see any obvious way to add anything new at the moment.

**Comment by nitsnatsnok** [comment]
@m-scarborough No, as a normal user that's currently not possible, and also not desired. We decided that for adding sources (as rarely as this should occur at this point) the user should contact any of the admins so that he can check the data and enter it correctly. I am at your disposal for that: konstantin.stephan@uni-jena.de

**Comment by PaulHeggarty** [comment]
Matthew is not a "normal user", he is one of just a handful of Indo-European cognacy experts and certainly should have rights to do this.  He is one of the most experienced members of the core CoBL team.  He does not need 'supervision' from an 'admin'.   He should have full rights to edit and delete existing entries, and to upload new bibtex records.  The same goes for Cassandra, Roland and Leszek, for a start, and many other experienced users.  It is the language specialists who know the sources best, and it is their responsibility to get them right.  The sources system just needs to be set up so that it is obvious how the data are to be entered correctly.

For now, I have amended Matthew's rights as per the screenshot below, but I suspect this will not work for the new bibtex upload system.  If so, we may need a new rights category for this.  And the other users I mention here should also have these rights.
![screenshot - 2016-11-07 20_16_33](https://cloud.githubusercontent.com/assets/8819910/20072028/20911b6a-a527-11e6-9bf0-7316082b1bb8.png)

**Comment by khoidt** [comment]
#### Update
Now in `devel`. The new version fixes the part of issue #344, introduces somewhat more elaborate views for source-related data and adds permission check for users who can edit or add or delete sources (instead of superuser status).
@runjak, deploy it, please.

**Comment by nitsnatsnok** [comment]
Maybe we misunderstood each other here – I tried to answer Matthew's general question whether adding sources is currently possible rather then question anybody's status here.

> The sources system just needs to be set up so that it is obvious how the data are to be entered correctly.

I don't know of any elaborate bibliography system which is useable intuitively, even userfriendly programs like Zotero need training to enter the fields correctly. The same goes for BibTeX, and as we use some additional/special fields it is cumbersome to get this correctly and completely. I will do my best to give explanations and instructions on maybe an own wiki page when the system is ready.

@khoidt:
- [ ] Please add a “Add Source” button with a mask to enter a new source to the list.

**Comment by PaulHeggarty** [comment]
You seem to be very much overcomplicating the task here.  CoBL is not, and has absolutely no aspiration to be, an 'elaborate bibliography system'.  We just need a fast, practical solution, using as much as possible of the system that has already been developed for CLLD, so that we don’t waste time on bespoke programming, when for our purposes all we need is a data table.  

You also overstate the need for 'training' here.  I never had any 'training' in Zotero or any other system.  I rarely have to enter any fields manually anyway, because they are already.  It is not rocket science.  We do not need a complex wiki page which probably nobody will look at because they consider that they know how to deal with a bibliographical reference.  If errors occur, we can fix them manually, very fast, so long as we have a suitable straightforward way to filter and edit the data table.

**Comment by nitsnatsnok** [comment]
OK, @khoidt let's concentrate on simply copying the CLLD system, see e.g. [WALS sources list](http://wals.info/refdb), [WALS source entry page](http://wals.info/refdb/record/Alwi-et-al-1998) and [WALS source](https://github.com/clld/wals3). Obviously they have an engine which roughly combines the data fields to form the “Text” line, this is exactly what we need rather than taking the citation_form field. A click on “more” should show this text line solely, also please add the short form (“Name” in WALS) column in the table. Making it clickable and open the record's page with at least the text line and the BibTeX form also seems like an idea worth adopting to me. This BibTeX output as well as the one downloadable as a file should not contain the following fields though:

- empty ones
- deprecated
- authortype
- TRS

**Comment by PaulHeggarty** [comment]
The new lookup system that now works is excellent, thanks.  
Some more, minor issues added above.  
Priorities are those issues listed in the first comment here.

**Comment by nitsnatsnok** [comment]
Here's a [.bib file](https://github.com/lingdb/CoBL/files/584012/corrected_names_01.txt) (with .txt extension of course) containing the revised author and editor names, they should all have the format requested [above](https://github.com/lingdb/CoBL/issues/344#issuecomment-258789827) now – with two exceptions:
- some authors/editors are institutions like "Pali Text Society" for ID 246, those are bracketed with **extra** {} as this is the common way to tell BibTeX that this is not a separable name
- some authors/editors (namely in the IDs 254 and 271) have a “and others” extension, since I was not able to find all authors/editors for every source – be careful to not render “others” as a name accidentally!

The following small things need to be adjusted on the output side yet:

.bib field(s)|.bib data|website's output
---|---|---
author |{Pali Text Society}|Pali Text Society
author, editor|and others|et al.
year|2007/2016|2007–2016

Of course I (or rather `sed`) could change this directly in the data, but as this is the BibTeX style required for a proper .bib export, I think we should change it only in the website's output – otherwise we'd have to change it back for the .bib export.

**Comment by khoidt** [comment]
### Update
Now in `devel`.
* Adds filter fields for `/source/`.
* Improves performance of queries on `Source`-related data.

@runjak, deploy it, please.

**Comment by nitsnatsnok** [comment]
@khoidt @runjak There seem to be some issues with the last update: I'm not able to reach the sources list, the website just gives “502 Bad Gateway”. Additionally, changes made on the source page (e.g. [here](http://cobl.info/source/47/edit/)) are not saved when the button „Submit“ is clicked (may be related).

**Comment by runjak** [comment]
Investigating…

**Comment by runjak** [comment]
For the `502` error on the sources list I think that a timeout could be possible.
I could visit the page a view minutes ago, but it took some time.
In addition when I'm viewing the times that the django debug toolbar reports, I get something like this:
![screenshot_2016-11-14_16-23-09](https://cloud.githubusercontent.com/assets/5417642/20270369/115a4f62-aa87-11e6-845c-1259b1034875.png)
![screenshot_2016-11-14_16-23-52](https://cloud.githubusercontent.com/assets/5417642/20270372/148c2354-aa87-11e6-8a6e-a7e946f875e6.png)
Which is already close to a minute.
There appears to be a rather big number of queries happening, with a given proportion of duplicates:
![screenshot_2016-11-14_16-24-49](https://cloud.githubusercontent.com/assets/5417642/20270405/38a026d2-aa87-11e6-94a3-05e63fb67cc2.png)
![screenshot_2016-11-14_16-25-01](https://cloud.githubusercontent.com/assets/5417642/20270409/3d366418-aa87-11e6-826b-3acb9fc4fa5c.png)

**Comment by runjak** [comment]
A look at the server logs confirms that the load times for something are too large:
![screenshot_2016-11-14_17-05-32](https://cloud.githubusercontent.com/assets/5417642/20271914/a0cef65c-aa8c-11e6-883d-1a3c8bf25f81.png)

My guess is that this would be the sources table.

**Comment by runjak** [comment]
The problems with editing sources is something that I cannot reproduce.
I've tried locally both with my account and with the account of @nitsnatsnok (by replacing password data in the database).

**Comment by nitsnatsnok** [comment]
I have no problems editing with the edit button in the sources list, but when I change something in the source view, e.g. http://cobl.info/source/141/edit/, click the submit button and close the page, it's not saved.

**Comment by runjak** [comment]
Yes - but if I edit a source the change shows up in the sources table for me.
Also when I edit the source again my change is still in there.

**Comment by nitsnatsnok** [comment]
OK, thanks for testing!

**Comment by nitsnatsnok** [comment]
@khoidt Please delete the source with ID 126 from the database. It's a double entry and I changed all remaining citations linked to it.

**Comment by PaulHeggarty** [comment]
Why does deleting a source need to be an issue on this thread?  Authorised users should be able to delete sources directly on the website.

**Comment by khoidt** [comment]
Strange. I checked it in the morning and again right now, and the page seems to open normally. In any case, I'll try to minimize the queries further: there is a problem with that indeed, thought the last update improves this with adding `select_related `and `prefetch_related` to the source-related data.

**Comment by nitsnatsnok** [comment]
@khoidt: Please make implementing delete and add functions top priority. Until I'm able to delete them myself, I keep track of sources to delete here:

ID|title|reason
---|---|---
126|Hittite Etymological Dictionary|duplicate
390|Luvian Lexical Notes|duplicate

**Comment by khoidt** [comment]
@nitsnatsnok, deleting sources should be available via `/source/(id)/delete/`. Did you try it?

**Comment by nitsnatsnok** [comment]
This works, thank you! I will take care of the above mentioned IDs myself.

**Comment by nitsnatsnok** [comment]
@khoidt The “502 Bad Gateway” error for the sources page is back, it worked the whole day for me until now.

**Comment by khoidt** [comment]
> Multiple authors should not have and repeated (A and B and C and D). 
> 	2 authors = A and B. 
> 	In short form, 3 or more = A et al. 
> 	Long form: A, B, C and D. 
> 	Also, do not use 'and others' in English, use only et al.

I queried the database (I have a ~2,5 weeks old dump) for multiple (more than one) " and " and single "and others" in the `shorthand` and `citation_text` fields of `Source` and haven't been able to find any. 
It normally goes A, B, C & D etc. @nitsnatsnok, did you refer to the `author`, `editor` etc. fields? If so, these fields have no long and short forms. I still can implement a migration to _unify_ the format in some single particular way, but if won't affect the `shorthand` and `citation_text` values, which are stored and not generated.

**Comment by nitsnatsnok** [comment]
I think Paul referred to what's displayed in the author field in the sources list. Did you see [my comment above](https://github.com/lingdb/CoBL/issues/344#issuecomment-259756779)? We can't change the bibtex input, because this would mess up the bibtex data, which we should preserve for the bibtex output. Also, we should not use `citation_text` for the long form any more, as the fields have to be editable and the long form needs to be computed from them (see [this comment](https://github.com/lingdb/CoBL/issues/344#issuecomment-259112418)) – therefore, `shorthand` can stay a field that's just stored (and I take it as my responsibility to make sure these fields are filled correctly), but the long form needs to be generated exactly like WALS does it.

**Comment by khoidt** [comment]
@nitsnatsnok, thanks for making this clear (once more!). However, I'm not sure if generating the `citation_text` instead of having a stored value is a good idea. Queries on sources are already very heavy, and this can make it much worse (I spent practically the whole week trying to improve performance, and it is still pretty poor). Perhaps, only if it's done on a button click for a single entry, and if that's really the only place where it is used. 
But then, again, implementing (and then debugging and perfecting) scripts to compute citations is, of course, doable but seems to cost much more time and effort then generating the titles with some proper citation software on the side and importing them to the database.
@runjak, @PaulHeggarty, what is your opinion?

**Comment by PaulHeggarty** [comment]
Do not waste time on "implementing (and then debugging and perfecting) scripts to compute citations".  We are not trying to programme a new bibliography manager software!  We just need a simple way for people to add a citation to their lexemes.  Forget everything else, and focus on getting that working.  We are months late on this.  Tomorrow is the start of our CoBL workshop and it is with huge embarassment that we will have to announce to them that they still cannot do so simple a thing as add a citation by just typing "Smith 1995", for example, in the 'citation cell' on any lexeme record row.  **That** is what about 70 people need, and have needed for months.

**Comment by nitsnatsnok** [comment]
[WALS](http://wals.info/refdb) just seems to combine the fields to form a simple text line, which is then displayed as the full citation – wouldn't it be a good idea to just copy this behaviour from them? If they compute it just once when the entry is added or edited, it doesn't need to eat up much performance.
The problem I see with stored values for the full citation line is that entries also have to be easily editable: The editor would need to not only edit the information in the fields correctly, but also adjust the citation line – this is cumbersome and prone to errors. However, in favor of a quick solution let's stay with `citation_text` for now.

**Comment by PaulHeggarty** [comment]
Since the sources reference table, and the desired functionality for it, seems to be sucking up inordinate amounts of time, please focus on a dedicated label field of the type Smith 1995, created manually if necessary to avoid wasting time trying to compute it, and use that as the field by which to sort and select citations when adding them to a lexeme.

**Comment by khoidt** [comment]
I'm on an autocomplete for Smith 1995. I understand that it is a top priority, so I'll do my very best to get it ready for deployment tonight.

**Comment by khoidt** [comment]
## Update
- Add autocomplete to `Source` citations in `lexeme_report` (add `django-autocomplete-light`).
_Note that this will not work in development, unless_ `django-debug-toolbar` _is disabled_.
- Add `shorthand` column to sources table.
- Fix table layout in filters row.
- Ignore in BibTeX output: `deprecated`, `authortype`, `TRS` and all fields with empty values.
- Restrict details preview to `citation_form` only.
- Replace 'and others' with 'et al.' in author and editor fields (`306_0133_and_others_to_et_al`).
---
Now in `devel` and ready for deployment (@runjak).

**Comment by runjak** [comment]
I've deployed the code, but http://www.cobl.info/sources/ gives me a `Bad Gateway` and takes long time to load at all for my local setup.

**Comment by nitsnatsnok** [comment]
For me too, but #354 is even worse!

**Comment by khoidt** [comment]
@runjak, could you roll back to the previous release?

**Comment by runjak** [comment]
Rolled back.

**Comment by khoidt** [comment]
## Update
Adds the following to the previous update:
- Fix performance issue in `/sources/`.
_Django queries on Source-related data are now overridden with raw SQL. Works like a charm!_
- Remove pagination in `/sources/`.
_Spliting the data to improve performance is no longer relevant._
- Fix table layout breaking when filtering sources.
- Remove `django_autocomplete` folder from `static` (see [this commentary](https://github.com/lingdb/CoBL/commit/71c37c0389d95834d6b48dc1f05cdf7cef335c2d#commitcomment-19920108)).
- Change Jquery script `src` for `django-autocomplete-light` in `lexeme_report.html` (see #354).
---
@runjak, for security reasons in `#306`, not `devel`. Will be ready for deployment when merged with an AMD fix for #354.

**Comment by runjak** [comment]
@khoidt I'm currently finishing #312, and will attend to fixing AMD and merging right after that. Hoping to get trough with this today (it looks possible).

**Comment by runjak** [comment]
@khoidt I've managed to arrange the AMD code so that the module inclusion and minification process works and the autocomplete feature is working on the `lexeme_report` too.

I'll test this before deployment and will send you an updated copy of the static directory.

To get make addition of JavaScript dependencies simpler for you it would probably be helpful to sit together and go trough the usage of `node, npm, grunt` and `bower` in the code. Once this is configured working with it is fairly comfortable.

**Comment by khoidt** [comment]
@runjak, great! I'd be indeed very grateful if you could make me a short introduction. Let me know (probably, better per e-mail) when you're available.

**Comment by nitsnatsnok** [comment]
> In short form, 3 or more = A et al.

I corrected the short forms according to this.

**Comment by khoidt** [comment]
## Update
- Add COinS metadata for Source model (collectable by Zotero, Citavi etc.). Update on #306, #344, closes #148.
---
In branch #306.

**Comment by m-scarborough** [comment]
Question: (I'd like to make sure I get this right so I don't break anything while attempting it.)  I'd like to add some references that I've just paraphrased in citing myself or within other citations while the citation system was down earlier.  As I understand what I have to do for this is to make .bib file with the bibliographical information in the correct BibTeX format, e.g. something like:

@incollection{Watkins 1982,
  author       = {Calvert Watkins}, 
  title        = {Notes on the Formations of the Hittite Neuter},
  booktitle    = {Investigationes Philologicae et Comparativae. Gedenkschrift für Heinz Kronasser},
  publisher    = {Harrassowitz Verlag},
  year         = 1982,
  editor       = {Erich Neu},
  pages        = {250-262},
}

And then upload that using the import feature at http://cobl.info/sources/import/ ???

**Comment by nitsnatsnok** [comment]
Thanks for asking! There's a simple way to add a source directly from the lexeme view now, “Add link to new Source +”:
![2016-11-28-102655_1366x768_scrot](https://cloud.githubusercontent.com/assets/2664162/20662831/6e1b9f66-b555-11e6-93fb-a9342e640f50.png)
By clicking the plus sign you get a mask where you can enter the bibliographical data, the system will then determine an ID for the source, which is used to refer to it internally – this is another thing than the shorthand “Watkins 1982”, which you can enter in the “shorthand” field for being used. Also, please enter all names in the format “surname, forename”. Pages ranges are given as “250--262” in the pages field.

**Comment by m-scarborough** [comment]
Ok, I'll try that and let you know how that goes.  Is it possible to have a link to this 'add new source' page put somewhere more visible in the sources page itself?

**Comment by m-scarborough** [comment]
Ok, when I tried to do that I got the page unavailable and the usual accompanying Russian warning about alcohol being the enemy of productivity.

<img width="405" alt="screen shot 2016-11-28 at 11 43 39" src="https://cloud.githubusercontent.com/assets/15671787/20667254/3c202fd0-b560-11e6-86a4-89d8e9445025.png">

Although!  I have checked the source-list and it *does* seem to be added now:

<img width="1210" alt="screen shot 2016-11-28 at 11 47 12" src="https://cloud.githubusercontent.com/assets/15671787/20667293/70d765cc-b560-11e6-9dfb-737023182bae.png">

**Comment by khoidt** [comment]
@m-scarborough and @nitsnatsnok, the import form should function well, so if there is a `.bib` file with new entries I see no reasons not to use it. If it misfunctions, please, open an issue.

**Comment by khoidt** [comment]
## Update
- Implement add button for adding new sources in `/sources/` (#306, #344).
- Add delete button Source for edit form in `/sources/` (#306, #344).
- Add `Bootstrap` tooltips for column headers "X" (deprecated) and "TRS" (traditional reference source) (#344).
- Add `Bootstrap` formatting to buttons.
- Fix layout in citation form (#357).
-----
@runjak, in #306, already merged with the most recent `devel`.

**Comment by nitsnatsnok** [comment]
### Delete Function ###
When it comes to [deleting sources](https://github.com/lingdb/CoBL/issues/366#issuecomment-263406364): There's clearly a feature missing to delete a source directly in the sources list! Deleting is currently possible by entering http://cobl.info/source/_ID_/delete, however this ID is displayed nowhere. One can get the ID from one of the count links, but for Adams 1999a there is none. (I can get the ID from one of my database dumps, but this doesn't solve the overall problem.)
- [ ] add a delete button in the “More” or “Edit” view – The table is wide enough and doesn't need this rarely used feature.

### Link Consistency ###
Cf.
cobl.info/sourc**es**/
cobl.info/sourc**es**/_ID_/cognacy/
cobl.info/sourc**es**/_ID_/cogset/
cobl.info/sourc**es**/_ID_/lexeme/
cobl.info/sourc**e**/_ID_
cobl.info/sourc**e**/_ID_/edit
cobl.info/sourc**e**/_ID_/delete
It would be only consistent to have sourc**es** in every of those URLs.
- [ ] change the links to be consistent

**Comment by nitsnatsnok** [comment]
> the import form should function well, so if there is a .bib file with new entries I see no reasons not to use it. If it malfunctions, please, open an issue.

@khoidt Which ID should one use as a key for a new entry in the bib file? E.g. `@book{entrykey,(…)}`

**Comment by khoidt** [comment]
@nitsnatsnok, New entries should be imported with **no ID** at all. If an existing ID is given, the DB entry will be rewritten.

**Comment by nitsnatsnok** [comment]
@khoidt What's the status of the `Multiple authors should not have and repeated` task here? Is this somehow related to the `name field split` task lastly discussed in #366?

**Comment by khoidt** [comment]
## Update
- Fix #368 (Import doesn't work).
- Fix #369 (Bibliography Fields appear too wide).
- Fix #373 (“Page unavailable” error in http://cobl.info/sources/).
- Work in progress on viewing/adding citation `/cognateclasslist/`: 
	- Added citation column with no. of citation and
	- a popup with details when clicked.
---
@runjak, in #306.

**Comment by khoidt** [comment]
## Update (d4db9f2)

* Make ENTRYTYPE field mandatory and provide it with a dropdown list of choices (#374).
* Fix Cancel button on cognate class citation form (#372).
* Filter Citation counts on Current + Jena 200 lists (#371).
---
@runjak, in #306.
